### ISSUE TITLE
How are dependent packages found when a local repo is used?

### ISSUE THREAD

**Issue by mhsimkin** [chocolatey/choco#1469]
Using Chocolatey Free edition.

Due to security restrictions, the majority of the servers I'm interested in using Chocolatey on do not have access to the internet.  I'm working on a PowerShell script to internalize a package.  However, there is some disconnect in my understanding of how to internalize a package.

As an example, the dotnet4.7.1 package is dependent on the KB2919355 package, which is dependent on another hotfix.  The installation process goes out and grabs the KB2919355 package from the chocolatey repo.

If I execute the chocolatey install command and specify a repo location, as below:
`choco install dotnet4.7.1 -s d:\chocorepo`
will the specified repo be searched for all dependencies?

If not, how can I force Chocolatey to only use a local repo that contains all the internalized packages?

Also, is there some choco config command that I can use to permanently change the repo location?  This way it doesn't always need to be passed on the install command line.

Thanks

-marc

**Comment by ferventcoder** [comment]
Hi @mhsimkin 

> Due to security restrictions, the majority of the servers I'm interested in using Chocolatey on do not have access to the internet. 

That's pretty normal with many of our customers. They set up internal repositories, remove the default community repository source, and set default sources to their internal sources (command at bottom).

> I'm working on a PowerShell script to internalize a package. However, there is some disconnect in my understanding of how to internalize a package.

Internalization is definitely recommended. To understand internalization (which is not simply downloading a package in most cases), it would be helpful to read over https://chocolatey.org/docs/how-to-recompile-packages.

> As an example, the dotnet4.7.1 package is dependent on the KB2919355 package, which is dependent on another hotfix. The installation process goes out and grabs the KB2919355 package from the chocolatey repo.
> If I execute the chocolatey install command and specify a repo location, as below:
> `choco install dotnet4.7.1 -s d:\chocorepo`
> will the specified repo be searched for all dependencies?

Yes, and only the specified source.

> Also, is there some choco config command that I can use to permanently change the repo location? This way it doesn't always need to be passed on the install command line.

Yes, see `choco source -?` or take a look at https://chocolatey.org/docs/commands-source

**Comment by ferventcoder** [comment]
Questions are best asked over the mailing list or Gitter (chat client). If you have followup questions, you can ask here, but we'd appreciate it if you follow the other channels. Thanks!

**Comment by mhsimkin** [comment]
I understand how repo sources can be added/remove/enabled/disabled via the choco sources command.

If I disable all the repos except my local one.  Will the choco install command attempt to find the dependent package on my local repo or will it fail?

Is there a defined search hierarchy for searching repos for packages? 

FYI, understood about using the mailing list going forward.

Thanks

-marc

**Comment by ferventcoder** [comment]
> If I disable all the repos except my local one. Will the choco install command attempt to find the dependent package on my local repo or will it fail?

It will only find packages and dependencies on specified sources of if sources are not explicitly specified, on the sources that are in the default set.

In other words:

* if you explicitly specify a source (or sources) on the command line, that is the ***only*** source that will be evaluated for a package and its dependencies.
* If you do not explicitly specify a source (or sources on the command line), then the only sources that will be evaluated for a package and its dependencies are the ones that are the default sources that are enabled.

> Is there a defined search hierarchy for searching repos for packages?

Yes, that is priority. Covered in the options/switches at https://chocolatey.org/docs/commands-source and `choco source -?`:

~~~
 --priority=VALUE
     Priority - The priority order of this source as compared to other 
       sources, lower is better. Defaults to 0 (no priority). All priorities 
       above 0 will be evaluated first, then zero-based values will be 
       evaluated in config file order. Available in 0.9.9.9+.
~~~

**Comment by ferventcoder** [comment]
fixed the wording a bit
